java/feathub-udf/flink-udf/src/main/java/com/alibaba/feathub/flink/udf/processfunction/SlidingWindowKeyedProcessFunction.java

```diff
@@ -96,93 +142,211 @@ public void onTimer(
                 }
-                descriptor.aggFunc.aggregate(curRow.getField(descriptor.inFieldName), rowTime);
-                hasRow = true;
+                Row curRow = state.timestampToRow.get(rowTime);
+                descriptor.aggFunc.retract(accumulator, curRow.getField(descriptor.inFieldName));
             }
+            leftIdxMap.put(descriptor.outFieldName, leftIdx);
+
+            outputRow.setField(descriptor.outFieldName, descriptor.aggFunc.getResult(accumulator));
         }
+        state.updateLeftTimestampIdx(leftIdxMap);
+        state.updateAccumulators(accumulators);
+        outputRow.setField(rowTimeFieldName, Instant.ofEpochMilli(timestamp));
 
-        state.pruneRow(
-                ctx.getCurrentKey(), timestamp - aggregationFieldsDescriptor.getMaxWindowSizeMs());
+        state.pruneRow(timestamp, aggregationFieldsDescriptor);
 
         if (!hasRow) {
-            // output nothing if no row is aggregated.
+            state.maxRegisteredTimer.clear();
+            if (expiredRowHandler == null) {
+                // output nothing if no row is aggregated.
+                return;
+            }
+            final Row lastRow = state.lastOutputRow.value();
+            if (lastRow == null) {
+                return;
+            }
+            expiredRowHandler.handleExpiredRow(out, lastRow, timestamp);
+            state.lastOutputRow.clear();
             return;
         }
 
-        final Row aggResultRow =
-                Row.of(
-                        aggregationFieldsDescriptor.getAggFieldDescriptors().stream()
-                                .map(d -> d.aggFunc.getResult())
-                                .toArray());
-
-        out.collect(
-                Row.join(
-                        ctx.getCurrentKey(),
-                        aggResultRow,
-                        Row.of(Instant.ofEpochMilli(timestamp))));
+        state.lastOutputRow.update(outputRow);
+        out.collect(outputRow);
     }
 
     /** The state of {@link SlidingWindowKeyedProcessFunction}. */
-    private static class MultiWindowSizeState {
+    private static class SlidingWindowState {
+
+        /** This MapState maps from row timestamp to the row. */
         private final MapState<Long, Row> timestampToRow;
 
-        // One KeyedProcessFunction can process rows from different keys. Unlike KeyedState, which
-        // is already scoped by key by Flink, we need to have a map that store the timestamp list
-        // for each key.
-        private final Map<Row, LinkedList<Long>> keyToTimestamps;
+        /**
+         * This ListState keeps all the row timestamp that has been added to the timestampToRow
+         * state. Since the operator assume rows are ordered by time, the timestamp list should be
+         * ordered as well.
+         */
+        private final ListState<Long> timestampList;
+
+        /**
+         * This ValueState keeps a map from output field name to the left index of the
```

Comment:
Can you define the semantics of `left index` in the doc?

For example, we can replace `left index of the timestampList` with this:

`index of the entry in the timestampList such that all entries before it must have been outside of the time window for this field`.